docs/rfcs/20201208_fragmentation.md

```diff
@@ -0,0 +1,303 @@
+
+### Modify the Allocator
+
+Another strategy would be to modify the Seastar allocator to improve fragmentation. 
+
+Per [this discussion](https://github.com/scylladb/seastar/issues/836#issuecomment-742924680) one small thing we can do reduce worst-case fragmentation is to increase the effective minimum span size to 16K. Currently, the large allocator only handles requests of at least that size already, but for some pool sizes the small allocator can request spans as small as 4K (but for many sizes it requests larger spans because of the _preferred size_ mechanism described above). Upping the minimum size of small page span requests would waste some memory in the small pool (due to partially filled spans), but overall this should _save_ memory since it would reduce the size of the pages array by ~75% per [Avi's comment](https://github.com/scylladb/seastar/issues/836#issuecomment-748531815): this would save perhaps 7M, much larger than the worst-case small pool waste of less than 1M.
+
+This won't "solve" fragmentation since there are still pages as small as 16K, but it is easy and seems to have no downside.
+
+More radical changes are possible. For example the allocator could reserve some fraction of the memory for each shard as a "huge pool". This wouldn't be used by the buddy allocator, but rather to satisfy allocations that might otherwise fail due to fragmentation. This allocator would be a more "traditional" `mmap`-based allocator which isn't restricted to a 1:1 relationship between VA and PA, and so doesn't suffer as much from physical memory fragmentation (since all allocations are multiples of 4K and 4K pages can be remapped freely by the OS). There are many variations here:
+
+ - Is the huge allocator used unconditionally for allocations above some size, or only when the buddy allocator fails?
+ - Is the huge pool per-shard or shared?
+ - Do we just use the native built-in libc allocator (via the saved malloc pointer) as the huge poor, or layer another allocator on top of the reserved storage?
+ - How exactly does the allocator reserve physical memory, since they mostly work in the virtual address domain? We need an allocator which carefully limits its physical memory use, but tries to use all of it e.g., by punching holes (e.g., with `MADV_DONTNEED`) in existing mappings.
+
+
+### Abandon the Allocator
+
+We know seastar works with the default system allocator: that's how we run it in debug mode (so we can use ASAN and friends). So in principle, we could use the system allocator, or some enhanced drop-in replacement instead. However, we would lose the significant benefits of the seastar allocator:
```

Comment:
Would a hybrid approach work in which if an allocation would otherwise fail in Seastar native allocator the allocator will complain loudly about the allocation but drop down to the default system allocator to satisfy the request?

Reply:
Yes, and this is sort of what I was getting at in the previous section starting at sentence "For example the allocator could reserve some fraction of the memory". I think of this as the "hybrid" strategy as you called it, but apparently I didn't use that word, oops!

I think you probably wouldn't want to use _exactly_ the system allocation w/o modification because then you're just exchanging OOMs for quasi-failure conditions like every allocation going to the system allocator which is using more and more memory, hitting swap etc (though of course swap & overcommit can be bounded). So I think you'd probably want at least total allocation size tracking so you can OOM when the bound is exceeded. There's sort of a spectrum for this fallback from almost exactly the system allocator (shared across shards, no reclaim, mmap lazily, less predictable OOM behavior) to basically like the seastar allocator backed by virtual memory (per shard, reclaim, maybe allocate-up-front, etc) all of which seem pretty interesting.

A question: what does "complain loudly" mean in production today?